### PR TITLE
test(Scrollbar): Backport new scrollbar implementation's tests

### PR DESCRIPTION
**Pull request by kdheepak**
This PR backports the tests from https://github.com/ratatui-org/ratatui/pull/847
The backported tests are made to pass with the current (i.e. old) implementation, so that a reference can be established for the changes in the new implementation.

### CODE REVIEW

File: src/widgets/scrollbar.rs
Comment (Valentin271) [inline src/widgets/scrollbar.rs:1137]: Instead of having generic cases and pass in an `assertion_message` (which I guess is only to debug which case failed), cases could be [named](https://docs.rs/rstest/latest/rstest/attr.rstest.html#optional-case-description).

EDIT: That's also a better description when the test succeeds.

File: src/widgets/scrollbar.rs
Comment (Valentin271) [inline src/widgets/scrollbar.rs:1172]: Maybe a comment here (and for the above test) to indicate why the scrollbar is completely filled. It confused me at first and it took me some time to figure it out.

File: src/widgets/scrollbar.rs
Comment (Valentin271) [inline src/widgets/scrollbar.rs:1201]: ~~Is there any point to this test? Testing that is renders empty instead of the block border?~~

Nit but it confused me, really this tests an `empty` bar and not an invisible bar (with no track or thumb).

File: src/widgets/scrollbar.rs
Comment (Valentin271) [inline src/widgets/scrollbar.rs:1372]: So here you're testing when there is a top and bottom bar right? Have you noticed weird behavior related to that?

**Review by Valentin271 — COMMENTED** [review]
Using named cases feels necessary to me. Other comments are just little improvements.

File: src/widgets/scrollbar.rs
Comment (joshka) [inline src/widgets/scrollbar.rs:1189]: extract the repetition to an rstest fixture:

```rust
#[fixture]
fn scrollbar_without_arrows() -> Scrollbar { ... }

fn render_scrollbar_simple(
        #[case] expected: &str,
        #[case] position: usize,
        #[case] content_length: usize,
        #[case] assertion_message: &str,
        scrollbar_without_arrows: Scrollbar,
    ) { ... }
```

The main benefit of this is that it gives the test data a consistent name, which represents that item.
Generally I'm not a fan of DRY in tests, but for exact equivalent test objects, I am definitely for it.

File: src/widgets/scrollbar.rs
Comment (kdheepak) [reply src/widgets/scrollbar.rs:1137]: If we use named tests this makes it uneven, e.g.:

image

Comment (joshka) [reply src/widgets/scrollbar.rs:1137]: yeah, the alignment is good to have for very visual style tests like this.

File: src/widgets/scrollbar.rs
Comment (kdheepak) [reply src/widgets/scrollbar.rs:1172]: For context this is what the old implementation does, so it is merely a "specification test". 

This entire PR is just to show how the new tests behave in the old implementation. I'm planning to revert the commit after this PR is merged to use the new tests.

Here's what the new tests do here:

image

File: src/widgets/scrollbar.rs
Comment (kdheepak) [reply src/widgets/scrollbar.rs:1189]: What do you think about merging this and then making improvements in the new PR? It'd be easier for me to not have to jump back and forth between branches.

File: src/widgets/scrollbar.rs
Comment (Valentin271) [reply src/widgets/scrollbar.rs:1137]: Hm I see ... too bad. Your call what you want to do.

File: src/widgets/scrollbar.rs
Comment (joshka) [reply src/widgets/scrollbar.rs:1172]: So is this perhaps due to the fact that content length and viewport are about the same size? If so, need to add more tests where that isn't the case.

File: src/widgets/scrollbar.rs
Comment (joshka) [reply src/widgets/scrollbar.rs:1201]: +1 comment / more tests around this.

File: src/widgets/scrollbar.rs
Comment (joshka) [reply src/widgets/scrollbar.rs:1372]: +1 this is confusing

File: src/widgets/scrollbar.rs
Comment (kdheepak) [reply src/widgets/scrollbar.rs:1189]: This old implementation showed full scrollbars whenever `content_length` was less than or equal to viewport area:

image

File: src/widgets/scrollbar.rs
Comment (joshka) [inline src/widgets/scrollbar.rs:720]: It's not obvious why this is rendered like this, so it's worth adding a longer comment for this test that documents what's going on here.

Additionally, it would more intuitive to read these tests as input (content_length) -> output (expected) rather than the reverse.

I don't quite understand the meaning of the name of this test. " test_renders_empty_with_content_length_is_zero"

Edit:
So, this is a test of the vertical rendering. It might be useful to use a vertical looking element for the bars, and have this represented as an array e.g. `[" #", " #", ...]`.
That said, the tests all have the same data except the first one. So it's could be better to split this into two tests rather than going for the visual approach.

File: src/widgets/scrollbar.rs
Comment (joshka) [inline src/widgets/scrollbar.rs:1143]: What does simplest mean here?
What concept does this test communicate?

File: src/widgets/scrollbar.rs
Comment (joshka) [inline src/widgets/scrollbar.rs:1178]: This name communicates very little about why this test is how it is.

**Review by joshka — CHANGES_REQUESTED** [review]
Generally, these tests need to be in a state where even if they are characterization tests, the character of the existing implementation is explained somehow.

This is sort of a [Chesterton's fence](https://fs.blog/chestertons-fence/) issue. i know you want to change this for the better, but I don't know what it currently does and why. The changes you want to make are a complete rewrite, so having tests that make it easy to see what's changed allows the reviewer of the rewrite to understand the changes better.

**Comment by kdheepak** [comment]
Closing in favor of original PR.

**Pull request closed without merging**
